### ISSUE TITLE
In the Assign Task flow, when typing an assignee’s email, there is massive lag

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#20075]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. In NewDot, press the green + button
2. Click Assign Task
3. Enter a title and click next
4. Click “Assignee”
5. Start typing a name or email
6. There is massive lag

## Expected result: 
When I enter a name or email in the Assignee screen, the name/email I type should appear in realtime as I type
## Actual result: 
The letters lag several seconds after I type them in the Assignee screen.
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:**
**Reproducible in staging?:** needs reproduction
**Reproducible in production?:** needs reproduction
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

**Expensify/Expensify Issue URL:**
**Issue reported by:** @sakluger 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1684539726401459

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by Thanos30** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Text Input lags when opening Assignee section on Task creation.

### What is the root cause of that problem?

The root cause of the issue lies on `TaskAssigneeSelectorModal.js` file, and specifically on the `updateOptions` function.

The `updateOptions` function is being called on `onChangeText` callback of the TextInput, and it is also being called on the second `useEffect` hook we have. This makes it re-render more than needed times, which causes the lag we face.

### What changes do you think we should make in order to solve the problem?
While not a big fan of the `setTimeout` function, in this case it is essential to use it, to make sure we don't call `updateOptions()` an unnecessary amount of times.

The second `useEffect` hook on `TaskAssigneeSelectorModal.js` file will look like this:
```
useEffect(() => {
        Timing.start(CONST.TIMING.SEARCH_RENDER);
        Performance.markStart(CONST.TIMING.SEARCH_RENDER);

        const debounceTimer = setTimeout(updateOptions, 300);

        return () => {
            clearTimeout(debounceTimer);
            Timing.end(CONST.TIMING.SEARCH_RENDER);
            Performance.markEnd(CONST.TIMING.SEARCH_RENDER);
        };
    }, [updateOptions]);

```
, where we set a debounceTimer for the call which we clear on return.

Video:

https://github.com/Expensify/App/assets/32620770/fe8cd065-9ff2-4498-9754-f794f59ad215

**Comment by francoisl** [comment]
Cool, thanks for the proposal @Thanos30. Removing the duplicate call to `updateOptions()` makes sense to me.

I'm less convinced about the `setTimeout`, do you still notice a performance issue without it? Also, is it easier than using `_.debounce()` in this situation?

**Comment by Thanos30** [comment]
@francoisl We can still work with only removing the duplicate call to `updateOptions()`, I just thought there is no need to trigger the search with every user type. We could possibly use the `_.debounce()` method too, it's up to you wether you would like to implement a debounce search here or not.

**Comment by francoisl** [comment]
I think we tend to prefer `_.debounce()` over `setTimeout()` in situations like this, but `_.debounce` also uses `setTimeout` under the hood, so I don't know if there's an advantage to using one over the other.

I'll test tomorrow on a workspace that has a large number of members, to see if there's a performance advantage to using `setTimeOut`/`_.debounce`.

In the meantime, @rushatgabhane do you have an opinion on this one?

**Comment by francoisl** [comment]
Ok, I ran more tests on a local domain with ~7500 accounts.

- On `main`:

https://github.com/Expensify/App/assets/2229301/99891650-07f3-45eb-bd5c-15c5717805f4

- Just removing the call to `updateOptions()` from `onChangeText`.  It feels better typing keys individually, but the lag is still visible if you smash a lot of keys very quickly

https://github.com/Expensify/App/assets/2229301/189b4592-cfea-4906-b34a-d808c98268ce

- With a `setTimeout()` and the delay set to 100 (300 feels too long). This one feels the smoothest IMO

https://github.com/Expensify/App/assets/2229301/baaef538-5ee9-4b69-bb84-7548a54fd157


Overall I think using the timeout might be the best improvement, what doo you think @rushatgabhane?

**Comment by Thanos30** [comment]
Did you try a combination of the 2? It smooths it out completely for me

**Comment by francoisl** [comment]
Yes, the last video I posted above is with a combination of the 2. Sorry if that wasn't clear, I'll edit my comment for clarity.

Were you testing on a [High Traffic account](https://github.com/Expensify/App/issues/20075) though? Let me know if you need to turn your account into a High Traffic account, or you can request it from the #expensify-open-source Slack channel also.

**Comment by Thanos30** [comment]
My apologies @francoisl , I misunderstood your words. I do use a High Traffic account, yeah. Maybe on a slightly bigger timeout the lag will completely vanish. On 300 timeout it was really smooth for me. We could do 150-200 maybe and see if it's better. I do get that 300 is too long though.

**Comment by sonialiap** [comment]
@francoisl does the suggestion to increase it to 150-200 sound good to you?

**Comment by francoisl** [comment]
Yeah though I'm still not sure if it's the cleanest solution. Would love to hear from @rushatgabhane or see if anyone else has a different proposal.

**Comment by Thanos30** [comment]
@rushatgabhane I think the reason this wasn't done inside `OptionsSelector` in the first place, is because the list of options is generated by each component using the `OptionsSelector` differently, based on the options available there (some may allow group chats, some don't etc).

We could possibly add a updateOptions() callback on the Options selector that we will pass via our components, and debounce run it when the value is changed. Let me know if you would like me to update my proposal on this.

**Comment by Thanos30** [comment]
@rushatgabhane @francoisl Component changed a little bit on a PR merged last week, but the problem still remains. I tried the method of making `getOptions()` on `OptionsListUtils.js` file, which is the method called to get all the available options with different parameters based on what results we want to get.

Basically we could handle the whole `getOptions()` function in a Promise with a setTimeout, so that this will be debounced every time. I concluded that this complicates things quite a bit, since we will always need to call that function asynchronously and get the result inside a .then(), and it's better to actually debounce that call per component, since this won't be a huge call every time.

Let me know what you both think on the above explanation, I can also update the Proposal now that the component changed a bit, but it's more or less the same fix.

**Comment by francoisl** [comment]
> Basically we could handle the whole getOptions() function in a Promise with a setTimeout, so that this will be debounced every time. I concluded that this complicates things quite a bit, since we will always need to call that function asynchronously and get the result inside a .then(), and it's better to actually debounce that call per component, since this won't be a huge call every time.

Not sure I follow - is there no way to use `_.debounce()` for this instead, without having to return a Promise?

**Comment by Thanos30** [comment]
@francoisl we are calling `getOptions()` from other functions to get the data we need. Since `_.debounce()` is used to run the function after a certain delay amount, our calling function will initially get an undefined result. We need to tell the calling function that it needs to wait for the debounce to be finished, and therefore we need to return a Promise and wait for it. 

Of course this is my assumption as I couldn't think of another way this would work.

**Comment by Thanos30** [comment]
@rushatgabhane That's a good point, but in our case I don't think this helps. The `leading`  feature makes the first time the function is called to not use a delay. This will bring back the issue we are facing, since it will immediately start looking for the available chat report to show when we open the assignees selection modal, which will cause the lag we want to remove.

Do you agree with that?

**Comment by Thanos30** [comment]
@rushatgabhane I understand. The reason it doesn't return `undefined` is because on the first call it doesn't use the delay, and therefore the promise immediately returns the value. In their example, they need to run their function immediately on the first call, and then they don't care for the rest, since they happen asynchronously and update a value.

In our case, we expect the result of the function to make use of it. I can run a couple of tests but this will not solve the lag issue when initially opening the modal.

Do you think it's better to have the debounce call on the `getOptions()` function, instead of debouncing that call on the components that use it?

**Comment by sonialiap** [comment]
@rushatgabhane bumping Thanos30's [question](https://github.com/Expensify/App/issues/20075#issuecomment-1608293919)

**Comment by Thanos30** [comment]
Right, this is my initial proposal too. What I mean is, we should debounce **calling** the function, and not actually **make** the function async and return a promise. The function doesn't actually do asynchronous tasks. So I believe it's better to debounce the call on the components that make use of the `getOptions` function, without changing the function itself.

**Comment by Thanos30** [comment]
Perfect, this brings us back to my initial proposal, debouncing the `getOptions` call. The code was altered at some point while we were discussing it and the duplicate call was removed, so I can just go with a refactor to debounce the call

**Comment by francoisl** [comment]
Yeah sounds good. Thanks for clarifying the proposal together 🙌

**Comment by Thanos30** [comment]
@rushatgabhane Forgot to ping here, PR is ready for review

**Comment by Thanos30** [comment]
@francoisl Since you were OOO yesterday and the PR was ready the day after the assignment, am I eligible for the bonus? 

Sorry for the inconvenience and thank you in advance!

**Comment by francoisl** [comment]
Yeah that would make sense to me, July 4<sup>th</sup> was a holiday (at least here in the US). @sonialiap is that something we can change for the final payment?

**Comment by Thanos30** [comment]
@francoisl  Thank you 🙏 , also, I read on Slack about a new automatic offer on Upwork, and I didn't apply for the job. Is that correct? or should I apply for the job?

**Comment by francoisl** [comment]
>  I read on Slack about a new automatic offer on Upwork, and I didn't apply for the job

I'm not aware of that, can you link to the Slack conversation where you saw that please? Otherwise perhaps @sonialiap you know what that would be about?

**Comment by Thanos30** [comment]
@francoisl https://expensify.slack.com/archives/C01GTK53T8Q/p1687969365336639

**Comment by Thanos30** [comment]
@francoisl HOLD for payment was on 2023-07-17. Bumping to make sure everything is okay

**Comment by francoisl** [comment]
Thanks for the reminder. I didn't see or hear about any issues with this, so let's go ahead with payments @sonialiap please.

**Comment by Thanos30** [comment]
Thank you @francoisl 🙏 Also, you probably missed it before, should I make an offer on the job after all? I didn't see the auto-assingment.

**Comment by francoisl** [comment]
Yes I think you'll need to, though I don't have access to Upwork myself to verify. @sonialiap can you assist please?

**Comment by sonialiap** [comment]
Sorry about the delay! I missed that the issue's title changed to "payment". Working on this now!

**Comment by Thanos30** [comment]
@sonialiap Thank you! Offer accepted

**Comment by sonialiap** [comment]
@rushatgabhane when you get a chance could you please work through [the checklist](https://github.com/Expensify/App/issues/20075#issuecomment-1629879838)

**Comment by rushatgabhane** [comment]
WIP ⚠️ 

1. The PR that introduced the bug has been identified. Link to the PR:

2. The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

3. A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

4. Determine if we should create a regression test for this bug.

5. If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again

**Comment by rushatgabhane** [comment]
Made a manual request on new dot https://staging.new.expensify.com/r/5352934870483260

**Comment by JmillsExpensify** [comment]
@sonialiap Can you please summarize the appropriate individual payments for all parties involved in this issue? This is holding up @rushatgabhane's NewDot payments. More information on this compliance process [in Slack](https://expensify.slack.com/archives/C01SKUP7QR0/p1690321157273869).

**Comment by sonialiap** [comment]
Issue reporter: none
Contributor: @Thanos30 $1500
Contributor+: @rushatgabhane $1500

**Comment by JmillsExpensify** [comment]
Reviewed details for @rushatgabhane. These details are accurate based on summary from Business Reviewer and are now approved for payment in NewDot.
